File: enode/src/main/java/org/enodeframework/messaging/impl/QueueMessageDispatching.java
package org.enodeframework.messaging.impl;

import org.enodeframework.messaging.Message;

import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

public class QueueMessageDispatching {
    private final DefaultMessageDispatcher dispatcher;
    private final RootDispatching rootDispatching;
    private final ConcurrentLinkedQueue<Message> messageQueue;

    public QueueMessageDispatching(DefaultMessageDispatcher dispatcher, RootDispatching rootDispatching, List<? extends Message> messages) {
        this.dispatcher = dispatcher;
        messageQueue = new ConcurrentLinkedQueue<>();
        messageQueue.addAll(messages);
        this.rootDispatching = rootDispatching;
        this.rootDispatching.addChildDispatching(this);
    }

    public Message dequeueMessage() {
        return messageQueue.poll();
    }

    public void onMessageHandled(Message message) {
        Message nextMessage = dequeueMessage();
        if (nextMessage == null) {
            rootDispatching.onChildDispatchingFinished(this);
            return;
        }
        dispatcher.dispatchSingleMessage(nextMessage, this);
    }
}

File: enode/src/main/java/org/enodeframework/messaging/impl/QueueMessageDispatching.kt
package org.enodeframework.messaging.impl

import org.enodeframework.messaging.Message
import java.util.concurrent.ConcurrentLinkedQueue

class QueueMessageDispatching(
    private val dispatcher: DefaultMessageDispatcher,
    rootDispatching: RootDispatching,
    messages: List<Message>
) {
    private val rootDispatching: RootDispatching
    private val messageQueue: ConcurrentLinkedQueue<Message> = ConcurrentLinkedQueue()

    init {
        messageQueue.addAll(messages)
        this.rootDispatching = rootDispatching
        this.rootDispatching.addChildDispatching(this)
    }

    fun dequeueMessage(): Message? {
        return messageQueue.poll()
    }

    fun onMessageHandled() {
        val nextMessage = dequeueMessage()
        if (nextMessage == null) {
            rootDispatching.onChildDispatchingFinished(this)
            return
        }
        dispatcher.dispatchSingleMessage(nextMessage, this)
    }
}
